### PR TITLE
New features: Allow specific BERT model and maximum edit distance value

### PR DESCRIPTION
**Pull request by dc-aichara**
**Scope of PR**
1. This PR solves 2 open issues. #15 #13 
1. Formatted code by black with code length 80.
1. Renamed few functions (These changes do not affect any functionality).

### CODE REVIEW

**Comment by R1j1t** [comment]
@dc-aichara thanks for the PR I will look at the changes in a few days and Merge it if everything seems fine. I do have some reservation on function/variable style changed from camelCase to snake_case.

File: contextualSpellCheck/contextualSpellCheck.py [hunk 1]
```diff
@@ -1,15 +1,33 @@
...
+
+    Returns:
+        List :If no suggestions: False else: True
+    """
+    return any(
+        [
+            token.i == suggestion.i and token.text == suggestion.text
+            for suggestion in token.doc._.suggestions_spellCheck.keys()
+        ]
+    )
```
Comment (R1j1t) [inline contextualSpellCheck/contextualSpellCheck.py:30]: How is this different from the original function at [L360 in master and L428 in this PR](https://github.com/R1j1t/contextualSpellCheck/blob/master/contextualSpellCheck/contextualSpellCheck.py#L360-L374)?

File: contextualSpellCheck/contextualSpellCheck.py [hunk 2]
```diff
@@ -73,121 +119,134 @@ def __init__(self, vocab_path="", debug=False, performance=False):
...
 
-            # {originalToken-1:[suggestedToken-1,suggestedToken-2,..],
-            #  originalToken-2:[...]}
             Doc.set_extension("suggestions_spellCheck", default={})
             Doc.set_extension("outcome_spellCheck", default="")
             Doc.set_extension("score_spellCheck", default=None)
 
             Span.set_extension(
-                "get_has_spellCheck", getter=self.span_require_spellCheck
+                "get_has_spellCheck", getter=self.span_require_spell_check
```
Comment (R1j1t) [inline contextualSpellCheck/contextualSpellCheck.py:150]: I thought about snake_case but I think as the name of the library is in CamelCase (**contextualSpellCheck**), I will like to stick with camelCase at least with "spellCheck". 
```suggestion
                "get_has_spellCheck", getter=self.spanRequireSpellCheck
```

or use function name styling with underscore:
```suggestion
                "get_has_spellCheck", getter=self.span_require_spellCheck
```
What I think is `spellCheck` should be left as is. I dont mind changing stles for others!

File: contextualSpellCheck/contextualSpellCheck.py [hunk 3]
```diff
@@ -1,15 +1,33 @@
 import spacy
 import torch
 import editdistance
-import datetime
+from datetime import datetime
```
Comment (R1j1t) [inline contextualSpellCheck/contextualSpellCheck.py:4]: A small but a good change

**Review by R1j1t — CHANGES_REQUESTED** [review]
The changes looks fine except a few, where I have added comment.

File: contextualSpellCheck/contextualSpellCheck.py (hunk 3 above)
Comment (dc-aichara) [reply contextualSpellCheck/contextualSpellCheck.py:4]: Thanks.

File: contextualSpellCheck/contextualSpellCheck.py (hunk 2 above)
Comment (dc-aichara) [reply contextualSpellCheck/contextualSpellCheck.py:150]: In my opinion, we use CamelCase  for class but for function snake_case is best.

File: contextualSpellCheck/contextualSpellCheck.py (hunk 1 above)
Comment (dc-aichara) [reply contextualSpellCheck/contextualSpellCheck.py:30]: Thank you for pointing out  this. That's my bad. I duplicated this function.

File: contextualSpellCheck/contextualSpellCheck.py [hunk 4]
```diff
@@ -454,35 +537,37 @@ def doc_suggestions_spellCheck(self, doc):
                 response[token].append(suggestion_score[0])
         return response
 
-    def doc_outcome_spellCheck(self, doc):
-        """Getter for Doc attribute
+    @staticmethod
+    def doc_outcome_spell_check(doc):
```
[the remark this answers is not in the document]
Comment (dc-aichara) [reply contextualSpellCheck/contextualSpellCheck.py]: Fixed

File: contextualSpellCheck/contextualSpellCheck.py (hunk 1 above)
Comment (dc-aichara) [reply contextualSpellCheck/contextualSpellCheck.py]: Removed duplicated function.

**Comment by dc-aichara** [comment]
@R1j1t This is weird that I see not `black` error on my local but it is failing while running checks. 
On my local:
```
~/Projects/contextualSpellCheck$ black . --check
All done! ✨ 🍰 ✨
8 files would be left unchanged.
```
In checks:
```

Run black . --check
would reformat /home/runner/work/contextualSpellCheck/contextualSpellCheck/contextualSpellCheck/contextualSpellCheck.py
would reformat /home/runner/work/contextualSpellCheck/contextualSpellCheck/examples/ja_example.py
would reformat /home/runner/work/contextualSpellCheck/contextualSpellCheck/contextualSpellCheck/tests/test_contextualSpellCheck.py
Oh no! 💥 💔 💥
3 files would be reformatted, 5 files would be left unchanged.
##[error]Process completed with exit code 1.

```
How about adding configuration file for black?

File: contextualSpellCheck/contextualSpellCheck.py (hunk 2 above)
Comment (R1j1t) [reply contextualSpellCheck/contextualSpellCheck.py:133]: I also read somewhere what in python snake_case preferred. Let go with that then, CamelCase for class and snake_case for function.

**Comment by R1j1t** [comment]
> How about adding [configuration file](https://black.readthedocs.io/en/stable/pyproject_toml.html#configuration-format) for black?

It is weird, on my local also the same issue. I re-run the job and it is failing again. This is strange. I am not sure how adding a config will help, because I am using the default `black` config and which  is also the case for GitHub actions. We also cannot skip the 3 files for which it is failing.

**Comment by dc-aichara** [comment]
@R1j1t It was black version issue. Black release new version 4 days back. 
To avoid version conflict, package version should be used.

**Comment by R1j1t** [comment]
Couple of things need to be updated:
1. Adding test cases for user specifying the transformer model and edit distance query 
2. At present, if I specify the `max_edit_dist=1, debug=True` in [L578](https://github.com/dc-aichara/contextualSpellCheck/blob/feat/specific_bert_model_support/contextualSpellCheck/contextualSpellCheck.py#L578), it gives an error.
```py
misspellings_dict[misspell] ['million', 'billion', ',', 'trillion', 'Million', '%', '##M', 'annually', '##B', 'USD']
Traceback (most recent call last):
  File "contextualSpellCheck/contextualSpellCheck.py", line 583, in <module>
    "Income was $9.4 milion compared to the prior year of $2.7 milion."
  File "/Users/rajat/miniconda3/lib/python3.7/site-packages/spacy/language.py", line 439, in __call__
    doc = proc(doc, **component_cfg.get(name, {}))
  File "contextualSpellCheck/contextualSpellCheck.py", line 170, in __call__
    self.candidate_ranking(doc, candidate)
  File "contextualSpellCheck/contextualSpellCheck.py", line 389, in candidate_ranking
    response[misspell],
KeyError: milion
```
3. At [L578](https://github.com/dc-aichara/contextualSpellCheck/blob/feat/specific_bert_model_support/contextualSpellCheck/contextualSpellCheck.py#L578) if I specify an edit distance where no candidate is selected (eg: `debug=False, max_edit_dist=1`), then 

[L590](https://github.com/dc-aichara/contextualSpellCheck/blob/feat/specific_bert_model_support/contextualSpellCheck/contextualSpellCheck.py#L590)
```py
print(doc._.performed_spellCheck)
# Current: True Expected: False, as no spellCheck was performed
```


[L597](https://github.com/dc-aichara/contextualSpellCheck/blob/feat/specific_bert_model_support/contextualSpellCheck/contextualSpellCheck.py#L597)
```py
doc[token_pos]._.get_require_spellCheck
# Current: False Expected: False
```

[605](https://github.com/dc-aichara/contextualSpellCheck/blob/feat/specific_bert_model_support/contextualSpellCheck/contextualSpellCheck.py#L605)
```py
print(doc[span_start:span_end]._.get_has_spellCheck)
# Current: False Expected: False

```

**Comment by dc-aichara** [comment]
> Adding test cases for user specifying the transformer model and edit distance query

**Do you mean Bert model?**

BTW in my experience `max_edit_distance` = 1 doesn't make sense for 2 words. We get edit distance 1 for 2 characters most of the time. 
Also, edit distance is not a good measure for non-English languages like Japanese Chinese etc. Strokes score provide better candidate ranking for these languages.

**Comment by R1j1t** [comment]
The reason for using edit distance 1 is to test how the code will behave if no suggestions are there. 

By test case for 
>test cases for user specifying the transformer 

I mean, if a user specifies a model which is not present, we should be able to handle that and also other test cases which would be useful. 

I completely agree that edit distance might not a good metric for language where word boundaries are not clear (like Chinese, Japanese(?) etc) . But I created the package keeping english in mind, and now I think as we are diversifying it would be good to incorporate it. So maybe for the time being, I will merge this PR (after you could add the test ). I will read about distance metric which can work for majority if not all languages and subsequently add the code. 

By the way, what is strokes Law? I could not find it in this [wikipedia list](https://en.wikipedia.org/wiki/String_metric#List_of_string_metrics).

**Comment by dc-aichara** [comment]
>I mean, if a user specifies a model which is not present, we should be able to handle that and also other test cases which would be useful.

I will add test for that. 

> But I created the package keeping English in mind

I got that. 

> By the way, what is strokes Law? 

Following links might be useful. 
1. [Japanese kanji strokes list ](https://en.wikipedia.org/wiki/List_of_kanji_by_stroke_count)
2. [kanjivg](https://github.com/KanjiVG/kanjivg)

**Comment by dc-aichara** [comment]
@R1j1t 
> if a user specifies a model which is not present, we should be able to handle that.


How do you want to handle this? Do you want to catch the error message from transformers when user specifies wrong model? Something like [this pytest](https://github.com/dc-aichara/contextualSpellCheck/blob/315f54ee3031a68e2a620fcac51d1c06f6ffb35f/contextualSpellCheck/tests/test_contextualSpellCheck.py#L545)  to make sure that `ContextualSpellCheck`  gives same error message always if model is not present?

**Comment by R1j1t** [comment]
Yes something similar will work, to inform that the model provided does not exist.

**Review by R1j1t — APPROVED** [review]

**Comment by R1j1t** [comment]
Thank you @dc-aichara for this great PR! The features you added were important to generalize this package. So thank you very much. I will add some extra test cases and then publish it on PyPI over the weekend. From your PR I also realized few things, which should be added to the repo, like:

- [x] black config
- [ ] flake8
- [ ] dependency version number 

If you have any suggestions for the package or the repository please feel free to let me know

**Comment by dc-aichara** [comment]
@R1j1t I am always happy to contribute to open source projects. :)

> flake8
minimum dependency version number

I can help with these .

I have invited you to [Gitter room](https://gitter.im/ContextaulSpellCheck/ContextualSpellCheck). Let's talk there.

**Comment by R1j1t** [comment]
I tried logging in the gitter room, but there some issue related to github access. 

But no worries. What I suggest is lets try to stick with github. I find some really amazing features which github has to offer. It would also make it easier for others to jump in and contribute! I have added some of my TODO features for the repo in this [kanban board](https://github.com/R1j1t/contextualSpellCheck/projects/3). Feel free to suggest extra features or any advice in general.

**Pull request merged**